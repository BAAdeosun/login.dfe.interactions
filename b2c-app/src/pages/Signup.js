import React from 'react';
import { animateScroll } from "react-scroll";

import components from '../components';
import { onChange } from '../helpers/pageUpdatesHandler';
import { PAGE_IDS } from '../constants/pageIds';

class Signup extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            firstName: null,
            lastName: null,
            email: null,
            password: null,
            tsAndCsAccepted: false,
            showErrors: false,
            showB2CErrors: true,
            errors: []
        }
        this.handleSubmit = this.handleSubmit.bind(this);
        this.onChange = onChange.bind(this);
    }

    handleSubmit(e) {
        e.preventDefault();
        //hide B2C errors and only show again if we are going to submit the form to B2C
        this.setState({ showB2CErrors: false });
        //update error messages
        this.state.errors.forEach((error) => {
            error.visible.text = error.current.text;
            error.visible.showSummaryText = error.current.showSummaryText;
        });
        //do something to validate and decide if we submit or show errors
        if (this.state.firstName &&
            this.state.lastName &&
            this.state.email &&
            this.state.password &&
            this.state.tsAndCsAccepted) {
            //hide our validation errors and prepare to show B2C ones (in case there are any)
            this.setState({ showErrors: false, showB2CErrors: true });
            //everything is valid, set data and submit B2C form
            this.setDataAndSubmit();
        }
        else {
            //show errors in each component
            this.setState({ showErrors: true });
            //scroll to the top of the page to show the errors
            animateScroll.scrollToTop({ duration: 500 });
        }
    }

    setDataAndSubmit() {
        //retrieve all elements we will need and set their values
        let b2cFirstName = document.getElementById('givenName');
        let b2cLastName = document.getElementById('surname');
        let b2cPassword = document.getElementById('newPassword');
        let b2cReenteredPassword = document.getElementById('reenterPassword');
        let b2cEmail = document.getElementById('email');
        let b2cTermsAndConditions = document.getElementById('tncCheckbox_true');
        let b2cSubmitButton = document.getElementById('continue');

        if (b2cFirstName && b2cLastName &&
            b2cPassword && b2cReenteredPassword &&
            b2cEmail &&
            b2cTermsAndConditions &&
            b2cSubmitButton) {
            b2cFirstName.value = this.state.firstName;
            b2cLastName.value = this.state.lastName;
            b2cPassword.value = this.state.password;
            b2cReenteredPassword.value = this.state.password;
            b2cEmail.value = this.state.email;
            b2cTermsAndConditions.checked = this.state.tsAndCsAccepted;
            //submit B2C form
            b2cSubmitButton.click();
        }
    }

    render() {

        const formContent =
            <div>
                <components.InputField
                    inputId='firstName'
                    inputLabel='First name'
                    onChange={this.onChange}
                    errorMessagePlaceholder='first name'
                    showErrors={this.state.showErrors}
                    errors={this.state.errors} />
                <components.InputField
                    inputId='lastName'
                    inputLabel='Last name'
                    onChange={this.onChange}
                    errorMessagePlaceholder='last name'
                    showErrors={this.state.showErrors}
                    errors={this.state.errors} />
                <components.InputField
                    type='email'
                    inputId='email'
                    inputLabel='Email address'
                    hint="You'll need this to sign in to your account"
                    onChange={this.onChange}
                    errorMessagePlaceholder='email address'
                    showErrors={this.state.showErrors}
                    errors={this.state.errors} />
                <components.CreateNewPassword
                    onChange={this.onChange}
                    showErrors={this.state.showErrors}
                    errors={this.state.errors} />
                <components.TermsAndConditions
                    onChange={this.onChange}
                    showErrors={this.state.showErrors}
                    errors={this.state.errors} />
            </div>

        const title = 'Create an account';

        const pageConfig = {
            title: title,
            header: title,
            formContent: formContent,
            submitButtonText: title,
            submitHandler: this.handleSubmit,
            errors: this.state.errors,
            showB2CErrors: this.state.showB2CErrors,
            errorSummaryContent: <components.PasswordHelp />
        };


        return (
            <div id={PAGE_IDS.SIGNUP}>
                <components.PageContainer pageConfig={pageConfig} />
            </div>
        )
    }
}

export default Signup;
