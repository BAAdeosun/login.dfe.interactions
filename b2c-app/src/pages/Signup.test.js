
import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Signup from './Signup';

it('renders without crashing', () => {
    shallow(<Signup />);
});

it('renders correctly', () => {
    const tree = renderer
        .create(<Signup />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});