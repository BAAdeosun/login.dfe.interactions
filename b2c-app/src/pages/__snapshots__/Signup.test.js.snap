// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  id="createAccount"
>
  <div>
    <div
      className="govuk-width-container"
    >
      <div
        className="govuk-breadcrumbs"
      >
        <ol
          className="govuk-breadcrumbs__list"
        >
          <li
            className="govuk-breadcrumbs__list-item"
          >
            <a
              className="govuk-breadcrumbs__link"
              href="https://nationalcareers.service.gov.uk/"
            >
              Home
            </a>
          </li>
        </ol>
      </div>
      <main
        className="govuk-main-wrapper"
      >
        <div
          className="govuk-grid-row"
        >
          <div
            className="govuk-grid-column-two-thirds"
          >
            <div
              className="pageLevelErrorContainer hide"
            >
              <div
                aria-labelledby="error-summary-title"
                className="govuk-error-summary"
                data-module="govuk-error-summary"
                role="alert"
                tabIndex="-1"
              >
                <h2
                  className="govuk-error-summary__title"
                  id="error-summary-title"
                >
                  There is a problem
                </h2>
                <div
                  className="govuk-error-summary__body"
                >
                  <ul
                    className="govuk-list govuk-error-summary__list"
                    id="errorSummaryItems"
                  />
                </div>
              </div>
            </div>
            <h1
              className="govuk-heading-xl govuk-!-margin-top-6 govuk-!-margin-bottom-9"
            >
              Create an account
            </h1>
            <form
              id="customForm"
              noValidate={true}
              onSubmit={[Function]}
            >
              <div>
                <div
                  className="govuk-form-group "
                >
                  <label
                    className="govuk-label"
                    htmlFor="firstNameCustom"
                  >
                    First name
                  </label>
                  <input
                    className="govuk-input govuk-!-width-one-half"
                    id="firstNameCustom"
                    name="firstName"
                    noValidate={true}
                    onChange={[Function]}
                    type="text"
                  />
                </div>
                <div
                  className="govuk-form-group "
                >
                  <label
                    className="govuk-label"
                    htmlFor="lastNameCustom"
                  >
                    Last name
                  </label>
                  <input
                    className="govuk-input govuk-!-width-one-half"
                    id="lastNameCustom"
                    name="lastName"
                    noValidate={true}
                    onChange={[Function]}
                    type="text"
                  />
                </div>
                <div
                  className="govuk-form-group "
                >
                  <label
                    className="govuk-label"
                    htmlFor="emailCustom"
                  >
                    Email address
                  </label>
                  <span
                    className="govuk-hint"
                    id="email-hint"
                  >
                    You'll need this to sign in to your account
                  </span>
                  <input
                    autoComplete="email"
                    className="govuk-input govuk-!-width-one-half"
                    id="emailCustom"
                    name="email"
                    noValidate={true}
                    onChange={[Function]}
                    spellCheck="false"
                    type="email"
                  />
                </div>
                <div>
                  <div
                    className="govuk-form-group "
                  >
                    <label
                      className="govuk-label"
                      htmlFor="newPasswordCustom"
                    >
                      Create new password
                    </label>
                    <input
                      className="govuk-input govuk-!-width-one-half"
                      id="newPasswordCustom"
                      name="newPassword"
                      noValidate={true}
                      onChange={[Function]}
                      type="password"
                    />
                  </div>
                  <div
                    className="govuk-form-group"
                  >
                    <details
                      className="govuk-details"
                    >
                      <summary
                        className="govuk-details__summary"
                      >
                        <span
                          className="govuk-details__summary-text"
                        >
                          Help choosing a valid password
                        </span>
                      </summary>
                      <div
                        className="govuk-details__text"
                      >
                        <div>
                          <p
                            className="govuk-body"
                          >
                            Your password must:
                          </p>
                          <ul
                            className="govuk-list govuk-list--bullet"
                          >
                            <li>
                              be between 8 and 16 characters
                            </li>
                            <li>
                              include at least one uppercase letter and one lowercase letter
                            </li>
                            <li>
                              include at least one number or one symbol
                            </li>
                          </ul>
                        </div>
                      </div>
                    </details>
                  </div>
                  <div
                    className="govuk-form-group "
                  >
                    <label
                      className="govuk-label"
                      htmlFor="reenteredPasswordCustom"
                    >
                      Re-type password
                    </label>
                    <input
                      className="govuk-input govuk-!-width-one-half"
                      id="reenteredPasswordCustom"
                      name="reenteredPassword"
                      noValidate={true}
                      onChange={[Function]}
                      type="password"
                    />
                  </div>
                </div>
                <div>
                  <h1
                    className="govuk-heading-m"
                  >
                    Terms and conditions
                  </h1>
                  <div
                    className="govuk-form-group "
                  >
                    <label
                      className="block-label"
                      htmlFor="tsAndCsCustom"
                    >
                      <input
                        aria-invalid="true"
                        checked={false}
                        id="tsAndCsCustom"
                        name="tsAndCsAccepted"
                        noValidate={true}
                        onChange={[Function]}
                        type="checkbox"
                        value={true}
                      />
                      I accept the 
                      <a
                        href="https://nationalcareers.service.gov.uk/help/terms-and-conditions"
                        id="tsAndCsLink"
                        rel="noopener noreferrer"
                        target="_blank"
                      >
                        terms and conditions
                      </a>
                       and I am 13 or over
                    </label>
                  </div>
                </div>
              </div>
              <button
                className="govuk-button"
                id="preSubmit"
                type="submit"
              >
                Create an account
              </button>
            </form>
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;
